session/session.go

```diff
@@ -1325,6 +1358,54 @@ func (s *session) Parse(ctx context.Context, sql string) ([]ast.StmtNode, error)
 	return stmts, nil
 }
 
+// ParseWithParams parses a query string, with arguments, to raw ast.StmtNode.
+func (s *session) ParseWithParams(ctx context.Context, sql string, args ...interface{}) ([]ast.StmtNode, error) {
+	sql, err := EscapeSQL(sql, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	internal := s.isInternal()
+
+	var stmts []ast.StmtNode
+	var warns []error
+	var parseStartTime time.Time
+	if internal {
+		// Do no respect the settings from clients, if it is for internal usage.
+		// Charsets from clients may give chance injections.
+		// Refer to https://stackoverflow.com/questions/5741187/sql-injection-that-gets-around-mysql-real-escape-string/12118602.
+		parseStartTime = time.Now()
+		stmts, warns, err = s.ParseSQL(ctx, sql, mysql.UTF8MB4Charset, mysql.UTF8MB4DefaultCollation)
```

Comment:
Is this a compatible change if a user upgrade the TiDB instance from an older version?

Reply:
Yes. Unless we write some internal SQLs that is using charsets incompatible with UTF8. Morgon said there should not be any UTF8 SQLs. I agreed with him.

Or arguments from users is not UTF8, which may become an injection attack. Do we allow user to create non-UTF8 username before? If not, it is safe.

Reply:
ok